godot-core/src/obj/script.rs

```diff
@@ -337,6 +337,33 @@ pub unsafe fn create_script_instance<T: ScriptInstance>(
     }
 }
 
+/// Checks if a script instance exists for a given object.
+///
+/// The engine does not expose the [`ScriptLanguage`] of a [`Script`] (as of 4.3) and it's therefore necessary to pass the expected
+/// language as well.
+///
+/// Use this function to implement [`IScriptExtension::instance_has`](crate::classes::IScriptExtension::instance_has).
+#[cfg(since_api = "4.2")]
+pub fn script_instance_exists<O, S, L>(object: &Gd<O>, script: &Gd<S>, language: &Gd<L>) -> bool
+where
+    O: Inherits<Object>,
+    S: Inherits<Script>,
+    L: Inherits<ScriptLanguage>,
+{
```

Comment:
For the special case of `ScriptExtension` (inherits `Script`), we have the virtual function [`_get_language()`](https://docs.godotengine.org/en/stable/classes/class_scriptextension.html#class-scriptextension-private-method-get-language).

And `Script` has 2 other inherited classes: `GDScript` + `CSharpScript`. For both of them, we know the language -- but I'm not sure if we have a `ScriptLanguage` instance for them.

Does this function even make sense for general `Script` objects, and not just `ScriptExtension` ones?

Or could we make the function more concrete, like:
```rs
pub fn script_instance_exists<O, S>(object: &Gd<O>, script: &Gd<S>) -> bool
where
    O: Inherits<Object>,
    S: Inherits<ScriptExtension>,
```

Reply:
Furthermore, you mention above

> Regardless of this, we sill should be able to use `script_instance_exists` in cases where we only have (obj, lang) by using `obj.get_script()` as the script.

So is `script` here equivalent to `object.get_script()`?
If yes, do we need the `script` parameter at all?

Reply:
> Does this function even make sense for general Script objects, and not just ScriptExtension ones?

Technically it would but on the other hand we have `Script::instance_has` that is implemented by the subclasses so I agree that we should be fine with just going with `S: Inherits<ScriptExtension>`.

> So is script here equivalent to object.get_script()?
> If yes, do we need the script parameter at all?

Not necessarely. In case of `IScriptExtension::instance_has(&self, object: Gd<Object>)`  we have an object and a script. One can always do `if object.get_script().to() == script { ... }` but for convenience I think it makes sense to include this check inside the function.

Reply:
But if the comparison is so simple, why is the implementation of the function so complex? Because you go through the C API instead of `Object::get_script()`?

I'm not against convenience, I just want to make sure I understand correctly :slightly_smiling_face: also because you initially mentioned

> After looking into #970 I noticed that there is currently no good way to implement `IScriptExtension::instance_has`.

Reply:
ah ok, so this function does two things:

1. check if `object.get_script()` is equal to `script`
2. check if `object_get_script_instance` returns a valid `GDExtensionScriptInstanceInfo` pointer for the `object` + `object.get_script().get_language()` pair.

Reply:
Ah, thanks! Could you add this to the function's docs?

Reply:
I expanded the description.